compiler/rustc_middle/src/ty/closure.rs

```diff
@@ -415,6 +416,52 @@ impl BorrowKind {
     }
 }
 
+pub fn analyze_coroutine_closure_captures<'a, 'tcx: 'a, T>(
+    parent_captures: impl IntoIterator<Item = &'a CapturedPlace<'tcx>>,
+    child_captures: impl IntoIterator<Item = &'a CapturedPlace<'tcx>>,
+    mut for_each: impl FnMut((usize, &'a CapturedPlace<'tcx>), (usize, &'a CapturedPlace<'tcx>)) -> T,
+) -> impl Iterator<Item = T> + Captures<'a> + Captures<'tcx> {
+    let mut parent_captures = parent_captures.into_iter().enumerate().peekable();
+    // Make sure we use every field at least once, b/c why are we capturing something
+    // if it's not used in the inner coroutine.
+    let mut field_used_at_least_once = false;
+    child_captures.into_iter().enumerate().map(move |(child_field_idx, child_capture)| {
+        loop {
+            let Some(&(parent_field_idx, parent_capture)) = parent_captures.peek() else {
+                bug!("we ran out of parent captures!")
+            };
+
+            let HirPlaceBase::Upvar(parent_base) = parent_capture.place.base else {
+                bug!("expected capture to be an upvar");
+            };
+            let HirPlaceBase::Upvar(child_base) = child_capture.place.base else {
+                bug!("expected capture to be an upvar");
+            };
+
+            if parent_base.var_path.hir_id != child_base.var_path.hir_id
+                || !std::iter::zip(
+                    &child_capture.place.projections,
+                    &parent_capture.place.projections,
+                )
+                .all(|(child, parent)| child.kind == parent.kind)
+            {
+                // Make sure the field was used at least once.
+                assert!(
+                    field_used_at_least_once,
+                    "we captured {parent_capture:#?} but it was not used in the child coroutine?"
+                );
+                field_used_at_least_once = false;
+                // Skip this field.
+                let _ = parent_captures.next().unwrap();
+                continue;
+            }
+
+            field_used_at_least_once = true;
+            break for_each((parent_field_idx, parent_capture), (child_field_idx, child_capture));
```

Comment:
The only thing this fails to do is assert that the parents captures are all used at least once. I'm tempted to use a coroutine for this and just use [`iter::from_coroutine`](https://doc.rust-lang.org/std/iter/fn.from_coroutine.html)...